### PR TITLE
osd: fix operator reconcile during OSD migration

### LINKED ISSUE #13525: Rook operator fails to keep track of the OSD migration process
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Operator fails to keep track of the OSD migration process. It needs to be restarted. It happens when the cluster has a lot of data and migration takes time for data backfill.

**Expected behavior:**
Operator should not require a restart. It should complete the migration of all the OSDs. 

**How to reproduce it (minimal and precise):**

1. Setup a rook cluster with OSDs on bluestore. The cluster should have a lot of data, say , 1 TB.
2. Update the cephCluster file with 
      ```
      store:
          type: bluestore-rdr
          updateStore: yes-really-update-store
       

3. Observe the operator logs. 

**File(s) to submit**:
[Uploading rook-ceph-operator.txt…]()



* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

**Logs to submit**:

* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary



  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:

* Output of kubectl commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook kubectl Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/kubectl-plugin)

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)):

### PR DESCRIPTION
**Pull request by sp98**
If migration of the OSDs is enabled, operator should wait for the pgs to be clean after the all the OSDs have reconciled. This will help add the exponential backoff for the operator restarts when OSD migration is needed but PGs are still rebalancing the data after last OSD migration.
Also, if the migration is pending at the end of the OSD reconcile, then trigger the reconcile again.

**Issue resolved by this Pull Request:**
Resolves #13525

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 1]
```diff
@@ -268,20 +268,30 @@ func (c *Cluster) Start() error {
 		return errors.Wrapf(err, "failed to update ceph storage status")
 	}
 
-	if c.replaceOSD != nil {
-		delOpts := &k8sutil.DeleteOptions{MustDelete: true, WaitOptions: k8sutil.WaitOptions{Wait: true}}
-		err := k8sutil.DeleteConfigMap(c.clusterInfo.Context, c.context.Clientset, OSDReplaceConfigName, namespace, delOpts)
-		if err != nil {
-			return errors.Wrapf(err, "failed to delete the %q configmap", OSDReplaceConfigName)
+	if c.spec.Storage.Store.UpdateStore == OSDStoreUpdateConfirmation {
+		if c.replaceOSD != nil {
```
Comment (subhamkrai) [inline pkg/operator/ceph/cluster/osd/osd.go:272]:
```suggestion
	if c.spec.Storage.Store.UpdateStore == OSDStoreUpdateConfirmation && c.replaceOSD != nil {
```

can we do this?

Comment (sp98) [reply pkg/operator/ceph/cluster/osd/osd.go:272]: Nope. The other checks inside the main if block are not related to `c.replaceOSD != nil` condition.

Comment (subhamkrai) [reply pkg/operator/ceph/cluster/osd/osd.go:272]: ohh I see it now, thanks

**Review by subhamkrai — APPROVED** [review]
lgtm

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 2]
```diff
@@ -268,20 +268,30 @@ func (c *Cluster) Start() error {
...
+			err := k8sutil.DeleteConfigMap(c.clusterInfo.Context, c.context.Clientset, OSDReplaceConfigName, namespace, delOpts)
+			if err != nil {
+				return errors.Wrapf(err, "failed to delete the %q configmap", OSDReplaceConfigName)
+			}
 		}
 
-		// Wait for PGs to be healthy before continuing the reconcile
-		_, err = c.waitForHealthyPGs()
+		// wait for the pgs to be healthy before attempting to migrate the next OSD
+		_, err := c.waitForHealthyPGs()
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/osd.go:281]: Would this be solved if we just returned an error if we failed to wait for healthy PGs? 
```suggestion
		pgsClean, err := c.waitForHealthyPGs()
		if !pgsClean {
			// return an error
		}
```

Comment (sp98) [reply pkg/operator/ceph/cluster/osd/osd.go:281]: Not really.  PGs will become healthy after the first OSD migration but we still need to reconcile again if the there are pending OSDs to be migrated. 

So two things are needed:
- Reconcile the  operator if there are pending OSDs for migration. 
- If the pgs are not healthy, then wait for 10 minutes for them get healthy before reconciling.

**Pull request merged**